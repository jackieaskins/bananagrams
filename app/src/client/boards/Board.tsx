import { Card, Flex } from "@chakra-ui/react";
import BoardSquare from "./BoardSquare";
import { convertToArray } from "./convert";
import { Board as BoardType } from "@/types/board";

type BoardProps = {
  board: BoardType;
};

export default function Board({ board }: BoardProps): JSX.Element {
  return (
    <Card display="inline-flex" flexDirection="column" variant="outline">
      {convertToArray(board).map((row, y) => (
        <Flex key={y}>
          {row.map((boardSquare, x) => (
            <BoardSquare key={x} x={x} y={y} boardSquare={boardSquare} />
          ))}
        </Flex>
      ))}
    </Card>
  );
}
