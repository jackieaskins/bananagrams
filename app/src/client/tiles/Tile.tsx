import { Box, Text } from "@chakra-ui/react";
import { useDrag } from "react-dnd";
import { Tile as TileType } from "@/types/tile";

type TileProps = {
  boardLocation: { x: number; y: number } | null;
  tile: TileType;
  color?: string;
};

export default function Tile({
  boardLocation,
  tile: { id, letter },
  color = "black",
}: TileProps): JSX.Element {
  const [{ isDragging }, dragRef] = useDrag({
    item: { type: "TILE", id, boardLocation },
    collect: (monitor) => ({
      isDragging: monitor.isDragging(),
    }),
  });

  const margin = boardLocation ? "0" : "5px";

  return (
    <Box
      ref={dragRef}
      alignItems="center"
      backgroundColor="yellow.100"
      border="1px"
      borderColor={color}
      borderRadius="5px"
      color={color}
      cursor="move"
      display="inline-flex"
      height="25px"
      justifyContent="center"
      margin={margin}
      opacity={isDragging ? 0.5 : 1}
      width="25px"
    >
      <Text as="b">{letter}</Text>
    </Box>
  );
}
