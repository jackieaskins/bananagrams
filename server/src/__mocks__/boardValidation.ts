export const validateAddTile = jest
  .fn()
  .mockImplementation((squares, { x, y }, tile) => {
    squares[x][y] = { tile };
    return squares;
  });

export const validateRemoveTile = jest
  .fn()
  .mockImplementation((squares, { x, y }) => {
    squares[x][y] = null;
    return squares;
  });
