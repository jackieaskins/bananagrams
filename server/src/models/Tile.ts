import BaseModel from './BaseModel';

export type TileJSON = {
  id: string;
  letter: string;
};

export default class Tile implements BaseModel<TileJSON> {
  private id: string;
  private letter: string;

  constructor(id: string, letter: string) {
    this.id = id;
    this.letter = letter;
  }

  getId(): string {
    return this.id;
  }

  getLetter(): string {
    return this.letter;
  }

  toJSON(): TileJSON {
    const { id, letter } = this;
    return { id, letter };
  }

  reset(): void {
    return;
  }
}
