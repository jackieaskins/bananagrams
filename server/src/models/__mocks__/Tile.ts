export default jest.fn().mockImplementation((id, letter) => ({
  getId: jest.fn().mockReturnValue(id),
  getLetter: jest.fn().mockReturnValue(letter),
  toJSON: jest.fn().mockReturnValue(`TileJSON - ${id} ${letter}`),
}));
