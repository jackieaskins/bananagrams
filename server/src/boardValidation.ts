import Dictionary from './dictionary/Dictionary';
import {
  BoardSquares,
  BoardPosition,
  Direction,
  BoardSquare,
  ValidationStatus,
  getSquareId,
} from './models/Board';
import Tile from './models/Tile';

type ValidatePosition = {
  start: BoardPosition;
  direction: Direction;
};
type BoardBoundaries = {
  rowStart: number;
  rowEnd: number;
  colStart: number;
  colEnd: number;
};

const getDelta = (direction: Direction): BoardPosition => {
  switch (direction) {
    case Direction.DOWN:
      return { row: 1, col: 0 };
    case Direction.ACROSS:
      return { row: 0, col: 1 };
  }
};

const getBoardBoundaries = (
  board: BoardSquares,
  { row, col }: BoardPosition
): BoardBoundaries => {
  let rowStart = row;
  let rowEnd = row;
  let colStart = col;
  let colEnd = col;

  Object.keys(board).forEach((id) => {
    const [r, c] = id.split(',');

    const row = parseInt(r);
    const col = parseInt(c);

    if (row < rowStart) rowStart = row;
    if (row > rowEnd) rowEnd = row;
    if (col < colStart) colStart = col;
    if (col > colEnd) colEnd = col;
  });

  return {
    rowStart,
    rowEnd,
    colStart,
    colEnd,
  };
};

const iterateWordFromStart = (
  board: BoardSquares,
  boardBoundaries: BoardBoundaries,
  start: BoardPosition,
  direction: Direction,
  loopFn: (square: BoardSquare) => boolean | void
): void => {
  const { rowEnd, colEnd } = boardBoundaries;
  const { row: rowDelta, col: colDelta } = getDelta(direction);
  let { row: nextRow, col: nextCol } = start;

  while (nextRow <= rowEnd && nextCol <= colEnd) {
    const square = board[getSquareId({ row: nextRow, col: nextCol })];
    if (!square || loopFn(square)) break;

    nextRow += rowDelta;
    nextCol += colDelta;
  }
};

const validateWordsAtPositions = (
  board: BoardSquares,
  boardBoundaries: BoardBoundaries,
  positionsToValidate: ValidatePosition[]
): void => {
  const positionsWithWords = positionsToValidate.map(({ start, direction }) => {
    const word: string[] = [];

    iterateWordFromStart(board, boardBoundaries, start, direction, (square) => {
      word.push(square.tile.getLetter());
    });

    return { start, direction, word: word.join('') };
  });

  positionsWithWords.forEach(({ start, direction, word }) => {
    let validationStatus: ValidationStatus;

    if (word.length === 1) {
      validationStatus = ValidationStatus.NOT_VALIDATED;
    } else {
      validationStatus = Dictionary.isWord(word.toLowerCase())
        ? ValidationStatus.VALID
        : ValidationStatus.INVALID;
    }

    iterateWordFromStart(board, boardBoundaries, start, direction, (square) => {
      square.wordInfo[direction].validation = validationStatus;
    });
  });
};

export const validateAddTile = (
  board: BoardSquares,
  position: BoardPosition,
  tile: Tile
): BoardSquares => {
  const boardBoundaries = getBoardBoundaries(board, position);

  const directions: Direction[] = Object.keys(Direction).map(
    (direction) => Direction[direction as keyof typeof Direction]
  );
  const { row, col } = position;

  const positionsToValidate: ValidatePosition[] = [];

  directions.forEach((direction) => {
    const { row: rowDelta, col: colDelta } = getDelta(direction);

    const beforeSquare =
      board[getSquareId({ row: row - rowDelta, col: col - colDelta })];
    const afterSquare =
      board[getSquareId({ row: row + rowDelta, col: col + colDelta })];

    const start = !!beforeSquare
      ? beforeSquare.wordInfo[direction].start
      : position;

    if (!!afterSquare) {
      iterateWordFromStart(
        board,
        boardBoundaries,
        { row: row + rowDelta, col: col + colDelta },
        direction,
        (square) => {
          square.wordInfo[direction].start = start;
        }
      );
    }

    positionsToValidate.push({ start, direction });
  });

  const wordInfo = positionsToValidate.map(({ direction, start }) => [
    direction,
    { start, validation: ValidationStatus.NOT_VALIDATED },
  ]);

  board[getSquareId({ row, col })] = {
    tile,
    wordInfo: Object.fromEntries(wordInfo),
  };

  validateWordsAtPositions(board, boardBoundaries, positionsToValidate);

  return board;
};

export const validateRemoveTile = (
  board: BoardSquares,
  position: BoardPosition
): BoardSquares => {
  const boardBoundaries = getBoardBoundaries(board, position);

  const directions: Direction[] = Object.keys(Direction).map(
    (direction) => Direction[direction as keyof typeof Direction]
  );
  const { row, col } = position;

  const positionsToValidate: ValidatePosition[] = [];
  directions.forEach((direction) => {
    const { row: rowDelta, col: colDelta } = getDelta(direction);

    const beforeSquare =
      board[getSquareId({ row: row - rowDelta, col: col - colDelta })];
    const afterSquare =
      board[getSquareId({ row: row + rowDelta, col: col + colDelta })];

    if (!!beforeSquare) {
      positionsToValidate.push({
        start: beforeSquare.wordInfo[direction].start,
        direction,
      });
    }

    if (!!afterSquare) {
      const newStart = { row: row + rowDelta, col: col + colDelta };
      iterateWordFromStart(
        board,
        boardBoundaries,
        newStart,
        direction,
        (square) => {
          square.wordInfo[direction].start = newStart;
        }
      );

      positionsToValidate.push({ start: newStart, direction });
    }
  });

  board[getSquareId({ row, col })] = null;

  validateWordsAtPositions(board, boardBoundaries, positionsToValidate);

  return board;
};
