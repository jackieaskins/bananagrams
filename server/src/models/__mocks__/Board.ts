import Tile from '../Tile';

export default jest.fn().mockImplementation(() => {
  const tiles: Record<string, Tile> = {};

  return {
    toJSON: jest.fn().mockReturnValue('BoardJSON'),
    reset: jest.fn(),
    validateEmptySquare: jest.fn(),
    removeTile: jest
      .fn()
      .mockImplementation((location) => tiles[`${location.x}${location.y}`]),
    addTile: jest.fn().mockImplementation((location, tile) => {
      tiles[`${location.x}${location.y}`] = tile;
    }),
  };
});
