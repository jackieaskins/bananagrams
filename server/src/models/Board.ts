import { validateAddTile, validateRemoveTile } from '../boardValidation';
import BaseModel from './BaseModel';
import Tile, { TileJSON } from './Tile';

export type BoardPosition = {
  row: number;
  col: number;
};
export enum Direction {
  ACROSS = 'ACROSS',
  DOWN = 'DOWN',
}
export enum ValidationStatus {
  NOT_VALIDATED = 'NOT_VALIDATED',
  VALID = 'VALID',
  INVALID = 'INVALID',
}
export type WordInfo = {
  start: BoardPosition;
  validation: ValidationStatus;
};
export type BoardSquare = {
  tile: Tile;
  wordInfo: Record<Direction, WordInfo>;
};
export type BoardRow = (BoardSquare | null)[];
export type BoardSquares = Record<string, BoardSquare | null>;
export type BoardJSON = Record<
  string,
  {
    tile: TileJSON;
    wordInfo: {
      [Direction.ACROSS]?: WordInfo;
      [Direction.DOWN]?: WordInfo;
    };
  } | null
>;

export const getSquareId = ({ row, col }: BoardPosition): string =>
  `${row},${col}`;

export default class Board implements BaseModel<BoardJSON> {
  private squares: BoardSquares = {};

  getSquares(): BoardSquares {
    return { ...this.squares };
  }

  getAllTiles(): Tile[] {
    return Object.values(this.squares)
      .filter((square) => square)
      .map((square) => (square as BoardSquare).tile);
  }

  reset(): void {
    this.squares = {};
  }

  clear(): Tile[] {
    const clearedTiles = this.getAllTiles();
    this.squares = {};
    return clearedTiles;
  }

  validateEmptySquare({ row, col }: BoardPosition): void {
    if (this.squares[getSquareId({ row, col })]) {
      throw new Error(`Board already has tile at position ${row}, ${col}`);
    }
  }

  toJSON(): BoardJSON {
    return Object.fromEntries(
      Object.entries(this.squares).map(([id, square]) => {
        if (square) {
          return [
            id,
            { tile: square.tile.toJSON(), wordInfo: square.wordInfo },
          ];
        }

        return [id, null];
      })
    );
  }

  removeTile({ row, col }: BoardPosition): Tile {
    const square = this.squares[getSquareId({ row, col })];
    if (!square) {
      throw new Error(`Board does not have a tile at position ${row}, ${col}`);
    }

    const { tile } = square;
    this.squares = validateRemoveTile(this.squares, { row, col });
    return tile;
  }

  addTile(position: BoardPosition, tile: Tile): void {
    this.validateEmptySquare(position);

    this.squares = validateAddTile(this.squares, position, tile);
  }
}
