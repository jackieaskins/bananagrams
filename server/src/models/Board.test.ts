import { validateAddTile, validateRemoveTile } from "../boardValidation";
import Board, { BoardSquares } from "./Board";
import Tile from "./Tile";

jest.mock("../boardValidation");

describe("Board Model", () => {
  let board: Board;

  beforeEach(() => {
    board = new Board();
  });

  describe("getSquares", () => {
    let squares: BoardSquares;

    beforeEach(() => {
      squares = board.getSquares();
    });

    test("initializes a 21 x 21 2-D array", () => {
      expect(squares).toHaveLength(21);
      expect(squares.every((row) => row.length === 21)).toBe(true);
    });

    test("makes all squares null by default", () => {
      expect(
        squares.every((row) => row.every((square) => square === null)),
      ).toBe(true);
    });
  });

  describe("getAllTiles", () => {
    test("returns empty array when no elements", () => {
      expect(board.getAllTiles()).toHaveLength(0);
    });

    test("returns tiles on board", () => {
      const tile = new Tile("A1", "A");
      board.addTile({ x: 0, y: 0 }, tile);

      expect(board.getAllTiles()).toEqual([tile]);
    });
  });

  describe("reset", () => {
    test("removes all tiles from board", () => {
      board.addTile({ x: 0, y: 0 }, new Tile("A1", "A"));
      board.reset();

      expect(board.getAllTiles()).toEqual([]);
    });
  });

  describe("clear", () => {
    const tile = new Tile("A1", "A");

    beforeEach(() => {
      board.addTile({ x: 0, y: 0 }, tile);
    });

    test("removes all tiles form board", () => {
      board.clear();

      expect(board.getAllTiles()).toEqual([]);
    });

    test("returns cleared tiles", () => {
      expect(board.clear()).toEqual([tile]);
    });
  });

  describe("validateEmptySquare", () => {
    const location = { x: 0, y: 0 };

    test("does not throw an error when location is empty", () => {
      expect(() => board.validateEmptySquare(location)).not.toThrow();
    });

    test("throws an error when location is not empty", () => {
      board.addTile(location, new Tile("A1", "A"));

      expect(() =>
        board.validateEmptySquare(location),
      ).toThrowErrorMatchingSnapshot();
    });
  });

  describe("toJSON", () => {
    test("turns fields into JSON", () => {
      board.addTile({ x: 0, y: 0 }, new Tile("A1", "A"));

      expect(board.toJSON()).toMatchSnapshot();
    });
  });

  describe("removeTile", () => {
    const location = { x: 0, y: 0 };
    const tile = new Tile("A1", "A");

    test("throws an error if there is no tile at location", () => {
      expect(() => board.removeTile(location)).toThrowErrorMatchingSnapshot();
    });

    test("updates squares to validated board after removing tile", () => {
      board.addTile(location, tile);
      board.removeTile(location);

      const squares = board.getSquares();

      expect(validateRemoveTile).toHaveBeenCalledWith(squares, location);
      expect(board.getSquares()[location.x][location.y]).toBeNull();
    });

    test("returns removed tile", () => {
      board.addTile(location, tile);

      expect(board.removeTile(location)).toEqual(tile);
    });
  });

  describe("addTile", () => {
    const location = { x: 0, y: 0 };
    const tile = new Tile("A1", "A");

    test("throws error if square is not empty", () => {
      board.addTile(location, tile);

      expect(() =>
        board.addTile(location, tile),
      ).toThrowErrorMatchingSnapshot();
    });

    test("updates squares to validated board after adding tile", () => {
      board.addTile(location, tile);

      const squares = board.getSquares();

      expect(validateAddTile).toHaveBeenCalledWith(squares, location, tile);
      expect(board.getSquares()[location.x][location.y]?.tile).toEqual(tile);
    });
  });
});
