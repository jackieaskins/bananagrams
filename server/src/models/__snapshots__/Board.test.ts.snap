// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Board Model addTile throws error if square is not empty 1`] = `"Board already has tile at position 0, 0"`;

exports[`Board Model removeTile throws an error if there is no tile at position 1`] = `"Board does not have a tile at position 0, 0"`;

exports[`Board Model toJSON turns fields into JSON 1`] = `
Object {
  "0,0": Object {
    "tile": Object {
      "id": "A1",
      "letter": "A",
    },
    "wordInfo": undefined,
  },
  "0,1": null,
}
`;

exports[`Board Model validateEmptySquare throws an error when position is not empty 1`] = `"Board already has tile at position 0, 0"`;
